code/modules/projectiles/projectile/beams.dm

```diff
@@ -185,3 +185,44 @@
 		var/mob/living/carbon/M = target
 		M.visible_message("<span class='danger'>[M] explodes into a shower of gibs!</span>")
 		M.gib()
+
+// a shrink ray that shrinks mobs, which grow back after a short while. shrunken mobs drop their stuff and move slower.
+/obj/item/projectile/beam/shrink
+	name = "shrink ray"
+	icon_state = "blue_laser"
+	hitsound = 'sound/weapons/shrink_hit.ogg'
+	damage = 0
+	damage_type = STAMINA
+	flag = "energy"
+	impact_effect_type = /obj/effect/temp_visual/impact_effect/shrink
+	light_color = LIGHT_COLOR_BLUE
+	var/shrink_time = 90
+
+/obj/item/projectile/beam/shrink/on_hit(atom/target, blocked = FALSE)
+	. = ..()
+	if(isopenturf(target))//shrunk floors wouldnt do anything except look weird
+		return
+	target.transform = target.transform.Scale(0.5,0.5)
+	addtimer(VARSET_CALLBACK(target, transform, target.transform.Scale(2,2)), shrink_time)
+	var/olddens = target.density
+	var/oldopac = target.opacity
+	target.density = 0
+	target.opacity = 0
+	addtimer(VARSET_CALLBACK(target, density, olddens), shrink_time)
+	addtimer(VARSET_CALLBACK(target, opacity, oldopac), shrink_time)

```

Comment:
This breaks if you shoot the gun at the same target in quick succession. This would probably fix that, but still isn't pretty.
```
addtimer(VARSET_CALLBACK(target, opacity, oldopac * target.opacity), shrink_time)
```
The more sane fix would be to apply some sort of trait to the object after it has been shrunk to prevent it from being shrunken again.

Reply:
Fixed with a component check